src/micro.c

```diff
@@ -0,0 +1,680 @@
+{
+    natsStatus s = NATS_OK;
+
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    IFOK(s, natsConn_getClosedCallback(&m->prev_on_connection_closed, &m->prev_on_connection_closed_closure, m->nc));
+    IFOK(s, natsConn_setClosedCallback(m->nc, on_connection_closed, m));
+
+    IFOK(s, natsConn_getDisconnectedCallback(&m->prev_on_connection_disconnected, &m->prev_on_connection_disconnected_closure, m->nc));
+    IFOK(s, natsConn_setDisconnectedCallback(m->nc, on_connection_disconnected, m));
+
+    IFOK(s, natsConn_getErrorCallback(&m->prev_on_error, &m->prev_on_error_closure, m->nc));
+    IFOK(s, natsConn_setErrorCallback(m->nc, on_error, m));
+
+    return microError_Wrapf(micro_ErrorFromStatus(s), "failed to wrap connection event callbacks");
+}
+
+static microError *
+unwrap_connection_event_callbacks(microService *m)
+{
+    natsStatus s = NATS_OK;
+
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    IFOK(s, natsConn_setClosedCallback(m->nc, m->prev_on_connection_closed, m->prev_on_connection_closed_closure));
+    IFOK(s, natsConn_setDisconnectedCallback(m->nc, m->prev_on_connection_disconnected, m->prev_on_connection_disconnected_closure));
+    IFOK(s, natsConn_setErrorCallback(m->nc, m->prev_on_error, m->prev_on_error_closure));
+
+    return microError_Wrapf(micro_ErrorFromStatus(s), "failed to unwrap connection event callbacks");
+}
+
+microError *
+microService_AddGroup(microGroup **new_group, microService *m, const char *prefix)
+{
+    if ((m == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(prefix) + 1); // "prefix.""
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    memcpy((*new_group)->prefix, prefix, strlen(prefix) + 1);
+    (*new_group)->m = m;
+    (*new_group)->next = m->groups;
+    m->groups = *new_group;
+
+    return NULL;
+}
+
+microError *
+microGroup_AddGroup(microGroup **new_group, microGroup *parent, const char *prefix)
+{
+    char *p;
+    int len;
+
+    if ((parent == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(parent->prefix) + 1 + // "parent_prefix."
+                                    strlen(prefix) + 1);         // "prefix\0"
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    p = (*new_group)->prefix;
+    len = strlen(parent->prefix);
+    memcpy(p, parent->prefix, len);
+    p[len] = '.';
+    p += len + 1;
+    memcpy(p, prefix, strlen(prefix) + 1);
+    (*new_group)->m = parent->m;
+    (*new_group)->next = parent->m->groups;
+    parent->m->groups = *new_group;
+
+    return NULL;
+}
+
+natsConnection *
+microService_GetConnection(microService *m)
+{
+    if (m == NULL)
+        return NULL;
+    return m->nc;
```

Comment:
Since this is a public API, don't we need locking to return m->nc? Or you guarantee that this is immutable?